crates/ra_vfs/src/io.rs

```diff
@@ -1,55 +1,184 @@
 use std::{
-    fmt,
-    fs,
+    fmt, fs,
     path::{Path, PathBuf},
+    sync::Arc,
+    thread,
 };
 
-use walkdir::{DirEntry, WalkDir};
-use thread_worker::{WorkerHandle};
+use crossbeam_channel::{Receiver, Sender};
+use parking_lot::Mutex;
 use relative_path::RelativePathBuf;
+use thread_worker::WorkerHandle;
+use walkdir::WalkDir;
 
-use crate::{VfsRoot, has_rs_extension};
+mod watcher;
+use watcher::Watcher;
+pub use watcher::WatcherChange;
 
-pub(crate) struct Task {
-    pub(crate) root: VfsRoot,
-    pub(crate) path: PathBuf,
-    pub(crate) filter: Box<Fn(&DirEntry) -> bool + Send>,
+use crate::{RootFilter, VfsRoot};
+
+pub(crate) enum Task {
+    AddRoot {
+        root: VfsRoot,
+        path: PathBuf,
+        root_filter: Arc<RootFilter>,
+        nested_roots: Vec<PathBuf>,
```

Comment:
I think `nested_roots` should just be a part of `RootFilter`, with a more generic name like `excluded_dirs`.